src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -1249,259 +414,43 @@ static void mongoSubCacheRefresh(const std::string& database)
+  BSONObj      update;
+  std::string  err;
 
-  //
-  // 1. Save subscriptionId, lastNotificationTime, and count for all items in cache
-  //
-  CachedSubscription* cSubP = mongoSubCache.head;
-
-  while (cSubP != NULL)
+  if (count != 0)
   {
-    CachedSubSaved* cssP       = new CachedSubSaved();
-
-    cssP->lastNotificationTime = cSubP->lastNotificationTime;
-    cssP->count                = cSubP->count;
+    // Update count
+    condition = BSON("_id"  << OID(subId));
+    update    = BSON("$inc" << BSON(CSUB_COUNT << count));
 
-    savedSubV[cSubP->subscriptionId]= cssP;
-    cSubP = cSubP->next;
-  }
-
-  LM_T(LmtMongoCacheSync, ("Pushed back %d items to savedSubV", savedSubV.size()));
-
-
-  //
-  // 2. Refresh cache (count set to 0)
-  //
-  mongoSubCacheRefresh();
-
-
-  //
-  // 3. Compare lastNotificationTime in savedSubV with the new cache-contents
-  //
-  cSubP = mongoSubCache.head;
-  while (cSubP != NULL)
-  {
-    CachedSubSaved* cssP = savedSubV[cSubP->subscriptionId];
-    if ((cssP != NULL) && (cssP->lastNotificationTime <= cSubP->lastNotificationTime))
+    if (collectionUpdate(collection, condition, update, false, &err) != true)
     {
-      // cssP->lastNotificationTime older than what's currently in DB => throw away
-      cssP->lastNotificationTime = 0;
+      LM_E(("Internal Error (error updating 'count' for a subscription)"));
     }
-    cSubP = cSubP->next;
   }
 
-
-  //
-  // 4. Update 'count' for each item in savedSubV where count != 0
-  // 5. Update 'lastNotificationTime' foreach item in savedSubV where lastNotificationTime != 0
-  //
-  cSubP = mongoSubCache.head;
-  while (cSubP != NULL)
+  if (lastNotificationTime != 0)
   {
-    CachedSubSaved* cssP = savedSubV[cSubP->subscriptionId];
-
-    if (cssP == NULL)
-    {
-      cSubP = cSubP->next;
-      continue;
-    }
-
-    std::string  tenant      = (cSubP->tenant == NULL)? "" : cSubP->tenant;
-    std::string  collection  = getSubscribeContextCollectionName(tenant);
-    BSONObj      condition;
-    BSONObj      update;
-    std::string  err;
-
-    if (cssP->count != 0)
-    {
-      // Update count
-      condition = BSON("_id" << OID(cSubP->subscriptionId));
-      update    = BSON("$inc" << BSON(CSUB_COUNT << cssP->count));
-    }
-    else if (cssP->lastNotificationTime != 0)
-    {
-      // Update lastNotificationTime
-      condition = BSON("_id" << OID(cSubP->subscriptionId) << CSUB_LASTNOTIFICATION << BSON("$lt" << cssP->lastNotificationTime));
-      update    = BSON("$set" << BSON(CSUB_LASTNOTIFICATION << cssP->lastNotificationTime));
-    }
-    else
-    {
-      cSubP = cSubP->next;
-      continue;
-    }
+    // Update lastNotificationTime
+    condition = BSON("_id" << OID(subId) << CSUB_LASTNOTIFICATION << BSON("$lt" << lastNotificationTime));
+    update    = BSON("$set" << BSON(CSUB_LASTNOTIFICATION << lastNotificationTime));
 
     if (collectionUpdate(collection, condition, update, false, &err) != true)
```

Comment:
"running in a separate process" -> "running in a separate block, one after other"